src/server.h

```diff
@@ -3420,6 +3420,15 @@ robj *setTypeDup(robj *o);
 #define HASH_SET_TAKE_VALUE (1 << 1)
 #define HASH_SET_COPY 0
 
+typedef struct {
+    sds value;
+    unsigned char field_offset;
+    char field_data[];
+} hashTypeEntry;
+hashTypeEntry *hashTypeCreateEntry(sds field, sds value);
+const void *hashTypeEntryGetKey(const void *entry);
+void freeHashTypeEntry(void *entry);
```

Comment:
Please make this one opaque and put it in `t_hash.c`.

Any accesses to `value` should use a function. Otherwise, we end up with lots of direct accesses to the value field and it it becomes very hard to implement embedded value in the future. (I believe embedded field+value can save considerable memory and CPU when it fits in one cache line, which should be a common case.)